### PR TITLE
net/lorawan : Support for IN865 region, Improve debugging

### PR DESCRIPTION
**Pull request by akshaim**
### Contribution description - 

This PR adds LoRaWAN support for IN865 Region. The PR also adds improved debugging.

### Testing procedure

- Tested gnrc_lorawan example in [examples](https://github.com/RIOT-OS/RIOT/tree/master/examples/gnrc_lorawan)

- Successfully tested for EU868 and IN865

#### EU868
Screenshot from 2020-09-24 16-15-04

#### IN865
IN865-Screen

### Issues/PRs references

~~OTAA in IN865 is not reliable, it works initially but fails often. Checked with RTL-SDR,  DUT ( b-l072z-lrwan1) seems to skip uplinks. ABP works flawlessly in EU868 and IN865 though.~~
Issues fixed in PR #15157

### CODE REVIEW

**Comment by jia200x** [comment]
> OTAA in IN865 is not reliable, it works initially but fails often. Checked with RTL-SDR, DUT ( b-l072z-lrwan1) seems to skip uplinks. ABP works flawlessly in EU868 and IN865 though.

This is really strange, since OTAA is only used for provisioning keys. There must be something else.
The code here doesn't touch these mechanisms, so I'm not sure what could be wrong there. Could you check if the stack is trying to send something to the radio?

File: sys/include/net/gnrc/lorawan/region.h [hunk 1]
```diff
@@ -26,12 +26,18 @@ extern "C" {
...
 static const uint32_t gnrc_lorawan_default_channels[] = {
+#ifdef LORAWAN_REGION_EU_868
     868100000UL,
     868300000UL,
     868500000UL
+#elif defined(LORAWAN_REGION_IN_865)
+    865062500UL,
+    865402500UL,
+    865985000UL
+#endif
```
Comment (jia200x) [inline sys/include/net/gnrc/lorawan/region.h:40]:
```suggestion
#if IS_ACTIVE(LORAWAN_REGION_EU_868)
    868100000UL,
    868300000UL,
    868500000UL
#elif IS_ACTIVE(LORAWAN_REGION_IN_865)
    865062500UL,
    865402500UL,
    865985000UL
#endif
```

File: sys/include/net/loramac.h [hunk 2]
```diff
@@ -102,6 +102,11 @@ extern "C" {
  * @name    LoRaMAC default values
  * @{
  */
+/**
+ * @brief Default LoRaWAN region
+ */
+#define LORAWAN_REGION_EU_868                  (true)
```
Comment (jia200x) [inline sys/include/net/loramac.h:108]: Note that the prefix of these functions here is `LORAMAC`. I would suggest to change this symbol to `LORAMAC_REGION_EU_868`

Comment (jia200x) [reply sys/include/net/loramac.h:108]: note that this hardcodes the region to EU 868. There should be guarded by an `#ifdef`

**Review by jia200x — CHANGES_REQUESTED** [review]
Thanks for this!
Here are some comments.

Also, could you please add the corresponding Kconfig file to the new symbols?

File: sys/include/net/loramac.h (hunk 2 above)
Comment (akshaim) [reply sys/include/net/loramac.h:108]: Definitely yes. Need your suggestion on the best possible method to so that we can implement additional channels as well. Should I have a symbol LORAMAC_REGION_ENABLE ?

Comment (jia200x) [reply sys/include/net/loramac.h:108]:
```suggestion
#if !IS_ACTIVE(LORAWAN_REGION_EU_868) && !IS_ACTIVE(LORAWAN_REGION_IN_865)
#define LORAWAN_REGION_EU_868                  (true)
#endif
```

**Comment by jia200x** [comment]
I just noticed there are some missing parts of the IN865 implementation (e.g CFList). Could you please cross check with the [regional specification document](https://lora-alliance.org/sites/default/files/2018-07/lorawan_regional_parameters_v1.0.3reva_0.pdf)?

File: sys/include/net/gnrc/lorawan/region.h (hunk 1 above)
Comment (akshaim) [reply sys/include/net/gnrc/lorawan/region.h:40]: This would only work once I move the macro to CONFIG_ namespace.

File: sys/include/net/loramac.h (hunk 2 above)
Comment (akshaim) [reply sys/include/net/loramac.h:108]: Okay. But the list could get a whole lot bigger once we add additional channels. I have added a macro `LORAMAC_REGION_ACTIVE` to take care of the same. Hope that works.

File: sys/include/net/gnrc/lorawan/region.h (hunk 1 above)
Comment (jia200x) [reply sys/include/net/gnrc/lorawan/region.h]: As commented above, I really think this change should introduce the proper Kconfig file

File: sys/include/net/loramac.h [hunk 3]
```diff
@@ -102,6 +102,13 @@ extern "C" {
  * @name    LoRaMAC default values
  * @{
  */
+/**
+ * @brief Default LoRaWAN region
+ */
+#ifndef LORAMAC_REGION_ACTIVE                  /**< To be used as Kconfig symbol */
```
Comment (jia200x) [inline sys/include/net/loramac.h:108]: As discussed offline, this macro represents something like "define a symbol EU868 if this macro is defined". This would ensure that one symbol is defined.

But IMO this is only a hack for the case where the user doesn't use Kconfig in order to ensure that there's always one symbol defined. Kconfig will always define at least one symbol.

For the non Kconfig case the only thing we need is to ensure that a default value is selected IF there's no symbol selected. In practice the user would simply define a region or leave the defaults there. Why would a user need to define "LORAMAC_REGION_ACTIVE" instead of simply defining EU868? (or skipping any configuration?)

As proposed below, I think this should be the way to go:
```
#if !IS_ACTIVE(LORAMAC_REGION_EU868) \
    && !IS_ACTIVE(LORAMAC_REGION_IN865 \
    && !IS_ACTIVE(LORAMAC_REGION_EU433) \
    ...
#define LORAMAC_REGION_EU868)
#endif
```

There are only 10 regions and they won't change in the short term. There's no code duplication and this list doesn't need special maintainance.

File: drivers/sx127x/sx127x_netdev.c [hunk 4]
```diff
@@ -46,6 +46,7 @@ void _on_dio3_irq(void *arg);
 
 static int _send(netdev_t *netdev, const iolist_t *iolist)
 {
+    DEBUG("[sx127x] Sending packet now.\n");
```
Comment (jia200x) [inline drivers/sx127x/sx127x_netdev.c:49]: the commit reflect a "lorawan" change, but this is a driver change. I'm ok with this piggyback commit, but at least it should have its own scope

File: sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mlme.c [hunk 5]
```diff
@@ -136,6 +136,7 @@ void gnrc_lorawan_mlme_process_join(gnrc_lorawan_t *mac, gnrc_pktsnip_t *pkt)
     gnrc_lorawan_process_cflist(mac, out + sizeof(lorawan_join_accept_t) - 1);
     mac->mlme.activation = MLME_ACTIVATION_OTAA;
     status = GNRC_LORAWAN_REQ_STATUS_SUCCESS;
+    puts("Successfully joined");
```
Comment (jia200x) [inline sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mlme.c:142]: This is a `sys` component. No puts should be there (otherwise this adds a dependency to stdio and adds ROM when in most cases is not necessary).

If you want to add a debug message, please add a DEBUG and add a prefix (e.g `DEBUG("[gnrc_lorawan]: Successfully joined!");`

**Comment by akshaim** [comment]
> I just noticed there are some missing parts of the IN865 implementation (e.g CFList). Could you please cross check with the [regional specification document](https://lora-alliance.org/sites/default/files/2018-07/lorawan_regional_parameters_v1.0.3reva_0.pdf)?

From what I have read through, following were found to be different :

1.  Data Rate : LoRa: SF7 / 250 kHz 11000 bits/s is not supported in IN865 compared to EU868
2.  TX Power . IN865 supports Max EIRP 16dB, 18dB and 20dB
3.  Receive windows : RX1DROffset varies.

File: sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mlme.c (hunk 5 above)
Comment (akshaim) [reply sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mlme.c:142]: Thanks for pointing this out. Even if debug is disabled shouldn't there be user feedback for successful join, similar to ones after successful uplinks ( confirmed ones) ?

File: sys/include/net/gnrc/lorawan/region.h (hunk 1 above)
Comment (akshaim) [reply sys/include/net/gnrc/lorawan/region.h]: Okay. I shall move them to "CONFIG_" namespace.

File: sys/include/net/loramac.h (hunk 3 above)
Comment (akshaim) [reply sys/include/net/loramac.h:108]: Considering non-Kconfig users, this makes sense.

File: drivers/sx127x/sx127x_netdev.c (hunk 4 above)
Comment (akshaim) [reply drivers/sx127x/sx127x_netdev.c:49]: Shall separate the commit :)

File: sys/include/net/loramac.h (hunk 3 above)
Comment (jia200x) [reply sys/include/net/loramac.h:108]: > Considering non-Kconfig users, this makes sense.

What makes sense? the original solution or the list?

File: sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mlme.c (hunk 5 above)
Comment (jia200x) [reply sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mlme.c:142]: Yes, there should be one. At that time there weren't too many mechanisms to inform an upper layer. I guess there's now the [Message Bus](https://doc.riot-os.org/msg__bus_8h.html). This could be used in a way that an upper layer can "subscribe" to a "Netif Up" event

**Comment by jia200x** [comment]
> From what I have read through, following were found to be different :
Data Rate : LoRa: SF7 / 250 kHz 11000 bits/s is not supported in IN865 compared to EU868
TX Power . IN865 supports Max EIRP 16dB, 18dB and 20dB
Receive windows : RX1DROffset varies.

Great! SF7 and 250 KHz is not implemented in RIOT anyway (yet). So that shouldn't be a problem.

File: sys/include/net/gnrc/lorawan/region.h (hunk 1 above)
Comment (jia200x) [reply sys/include/net/gnrc/lorawan/region.h]: yes please, and a proper Kconfig file :)

**Comment by jia200x** [comment]
nit-pick: regarding commit names, please separate the scope of those who belong to the driver (`drivers/sx127x: ...`), those who touch the common definitions of LoRaWAN (`net/lorawan`) and GNRC LoRaWAN specific commits (`gnrc/lorawan: ...`)

**Comment by jia200x** [comment]
I would say this draft can be migrated to PR

File: sys/include/net/loramac.h (hunk 3 above)
Comment (akshaim) [reply sys/include/net/loramac.h:108]: I meant the list you suggested makes sense for a non-Kconfig user

Comment (jia200x) [reply sys/include/net/loramac.h:108]: ah, got it

File: sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mlme.c (hunk 5 above)
Comment (akshaim) [reply sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mlme.c:142]: As you had suggested, I shall use a separate PR for the same.

Comment (jia200x) [reply sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mlme.c:142]: Yes. But as discussed offline, that should be a generic GNRC netif component.

File: sys/include/net/gnrc/lorawan/region.h (hunk 1 above)
Comment (akshaim) [reply sys/include/net/gnrc/lorawan/region.h]: Don't you think it's better to do Kconfig exposure on a separate PR ?

Comment (jia200x) [reply sys/include/net/gnrc/lorawan/region.h]: since there's an ongoing migration to Kconfig, this should at least include the "CONFIG_" prefix. Otherwise there will be an immediate follow up renaming these macros anyway

File: sys/include/net/loramac.h [hunk 6]
```diff
@@ -102,6 +102,13 @@ extern "C" {
  * @name    LoRaMAC default values
  * @{
  */
+/**
+ * @brief Default LoRaWAN region
+ */
+#if !defined(LORAMAC_REGION_EU868) && !defined(LORAMAC_REGION_IN865)
+#define LORAMAC_REGION_EU_868                  (1)
```
Comment (jia200x) [inline sys/include/net/loramac.h:109]:
```suggestion
#define LORAMAC_REGION_EU_868                  1
```

File: sys/include/net/loramac.h [hunk 7]
```diff
@@ -102,6 +102,13 @@ extern "C" {
  * @name    LoRaMAC default values
  * @{
  */
+/**
+ * @brief Default LoRaWAN region
+ */
+#if !defined(LORAMAC_REGION_EU868) && !defined(LORAMAC_REGION_IN865)
```
Comment (jia200x) [inline sys/include/net/loramac.h:108]:
```suggestion
#if !defined(LORAMAC_REGION_EU_868) && !defined(LORAMAC_REGION_IN_865)
```

File: sys/include/net/loramac.h [hunk 8]
```diff
@@ -203,14 +210,22 @@ extern "C" {
  * @brief   Default second RX window datarate index
  */
 #ifndef LORAMAC_DEFAULT_RX2_DR
+#ifdef LORAMAC_REGION_EU_868
 #define LORAMAC_DEFAULT_RX2_DR                 (LORAMAC_DR_0)
+#elif defined(LORAMAC_REGION_IN_865)
```
Comment (jia200x) [inline sys/include/net/loramac.h:215]: Missing IS_ACTIVE here.

Also as described in another comment, this should be at least prefixed with "CONFIG_"

File: sys/net/gnrc/link_layer/lorawan/gnrc_lorawan.c [hunk 9]
```diff
@@ -132,6 +132,10 @@ static void _config_radio(gnrc_lorawan_t *mac, uint32_t channel_freq, uint8_t dr
 
 static void _configure_rx_window(gnrc_lorawan_t *mac, uint32_t channel_freq, uint8_t dr)
 {
+    if (channel_freq) {
+        DEBUG("gnrc_lorawan: RX2 Frequency: %luHz \n", channel_freq);
+    }
```
Comment (jia200x) [inline sys/net/gnrc/link_layer/lorawan/gnrc_lorawan.c:137]:
```suggestion
    DEBUG("gnrc_lorawan: RX2 Frequency: %luHz \n", channel_freq);
```

The DEBUG macro already optimizes out the `channel_freq` variable

File: sys/include/net/gnrc/lorawan/region.h (hunk 1 above)
Comment (akshaim) [reply sys/include/net/gnrc/lorawan/region.h]: Done

File: sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mlme.c (hunk 5 above)
Comment (akshaim) [reply sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mlme.c]: Yup :+1:

File: sys/net/gnrc/link_layer/lorawan/gnrc_lorawan.c [hunk 10]
```diff
@@ -169,6 +173,7 @@ void gnrc_lorawan_event_tx_complete(gnrc_lorawan_t *mac)
 void gnrc_lorawan_event_timeout(gnrc_lorawan_t *mac)
 {
     (void) mac;
+    DEBUG("gnrc_lorawan: RX1 timeout.\n");
```
Comment (jia200x) [inline sys/net/gnrc/link_layer/lorawan/gnrc_lorawan.c:176]: this DEBUG is misplaced. It should be inside the switch

File: examples/gnrc_lorawan/README.md [hunk 11]
```diff
@@ -28,7 +28,11 @@ ifconfig 3 set appeui BBBBBBBBBBBBBBBB
 ifconfig 3 set appkey CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC
 ifconfig 3 up
 ```
+If chirpstack is being used, set appeui as follows
```
Comment (jia200x) [inline examples/gnrc_lorawan/README.md:31]:
```suggestion
If Chirpstack is being used, set AppEUI as follows
```

**Comment by jia200x** [comment]
I made a last round of comments. Overall it looks good.
If these comments get addressed I'm OK to move one with this one.

File: sys/net/gnrc/link_layer/lorawan/gnrc_lorawan.c [hunk 12]
```diff
@@ -132,6 +132,8 @@ static void _config_radio(gnrc_lorawan_t *mac, uint32_t channel_freq, uint8_t dr
 
 static void _configure_rx_window(gnrc_lorawan_t *mac, uint32_t channel_freq, uint8_t dr)
 {
+    DEBUG("gnrc_lorawan: RX2 Frequency: %luHz \n", channel_freq);
```
Comment (jia200x) [inline sys/net/gnrc/link_layer/lorawan/gnrc_lorawan.c:135]: Same thing here, this is called for RX1 and RX2

**Comment by jia200x** [comment]
> From what I have read through, following were found to be different :
    Data Rate : LoRa: SF7 / 250 kHz 11000 bits/s is not supported in IN865 compared to EU868
    TX Power . IN865 supports Max EIRP 16dB, 18dB and 20dB
    Receive windows : RX1DROffset varies.

I forgot this too. That has to be addressed in order to be compliant with the standard

File: sys/include/net/loramac.h (hunk 6 above)
Comment (aabadie) [reply sys/include/net/loramac.h]: Sorry, to jump-in late, but I think there define name should rather be `LORAMAC_REGION_EU868` (no `_` between EU and 868). This notation is used everywhere in the lorawan specs.

I haven't checked in detail what the rationale between the addition of this `_`, maybe there's a good reason ?

File: examples/gnrc_lorawan/README.md [hunk 13]
```diff
@@ -28,7 +28,11 @@ ifconfig 3 set appeui BBBBBBBBBBBBBBBB
 ifconfig 3 set appkey CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC
 ifconfig 3 up
 ```
+If Chirpstack is being used, set AppEUI as follows
```
Comment (aabadie) [inline examples/gnrc_lorawan/README.md:31]: I don't understand why using a Chirpstack server would need to clear the appeui.

Comment (jia200x) [reply examples/gnrc_lorawan/README.md:31]: Actually it's not necessary to clear it. It just the fact that the default Application Server of Chirpstack ignores the AppEUI

Comment (akshaim) [reply examples/gnrc_lorawan/README.md:31]: Yup as @jia200x  mentioned. There is no real purpose hence zeroed.

File: sys/include/net/loramac.h (hunk 6 above)
Comment (akshaim) [reply sys/include/net/loramac.h]: Dash helps with readability to identify the frequency `EU_868`, at least for me. Alliance do not use '_' though.

File: examples/gnrc_lorawan/README.md (hunk 13 above)
Comment (jia200x) [reply examples/gnrc_lorawan/README.md:31]: maybe this should be reflected differently:
E.g `If Chirpstack is being used, the AppEUI is ignored`

File: sys/net/gnrc/link_layer/lorawan/gnrc_lorawan.c [hunk 14]
```diff
@@ -171,13 +171,13 @@ void gnrc_lorawan_event_tx_complete(gnrc_lorawan_t *mac)
...
 {
     (void) mac;
-    DEBUG("gnrc_lorawan: RX1 timeout.\n");
     switch (mac->state) {
         case LORAWAN_STATE_RX_1:
             _configure_rx_window(mac, LORAMAC_DEFAULT_RX2_FREQ, mac->dl_settings & GNRC_LORAWAN_DL_RX2_DR_MASK);
             mac->state = LORAWAN_STATE_RX_2;
             break;
         case LORAWAN_STATE_RX_2:
+            DEBUG("gnrc_lorawan: RX1 timeout.\n");
```
Comment (jia200x) [inline sys/net/gnrc/link_layer/lorawan/gnrc_lorawan.c:180]:
```suggestion
            DEBUG("gnrc_lorawan: RX2 timeout.\n");
```

File: sys/net/gnrc/link_layer/lorawan/gnrc_lorawan.c [hunk 15]
```diff
@@ -176,7 +176,7 @@ void gnrc_lorawan_event_timeout(gnrc_lorawan_t *mac)
             mac->state = LORAWAN_STATE_RX_2;
             break;
         case LORAWAN_STATE_RX_2:
-            DEBUG("gnrc_lorawan: RX1 timeout.\n");
+            DEBUG("gnrc_lorawan: RX2 timeout.\n");
```
Comment (jia200x) [inline sys/net/gnrc/link_layer/lorawan/gnrc_lorawan.c:180]: Also please don't forget a message for the RX1 timeout :)

File: sys/net/gnrc/link_layer/lorawan/gnrc_lorawan.c (hunk 12 above)
Comment (akshaim) [reply sys/net/gnrc/link_layer/lorawan/gnrc_lorawan.c]: It was supposed to be RX frequency. Updated.

File: sys/net/gnrc/link_layer/lorawan/gnrc_lorawan.c (hunk 15 above)
Comment (akshaim) [reply sys/net/gnrc/link_layer/lorawan/gnrc_lorawan.c:180]: Already done :+1:

File: sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_region.c [hunk 16]
```diff
@@ -44,7 +45,16 @@ int gnrc_lorawan_set_dr(gnrc_lorawan_t *mac, uint8_t datarate)
 
 uint8_t gnrc_lorawan_rx1_get_dr_offset(uint8_t dr_up, uint8_t dr_offset)
 {
+#if (IS_ACTIVE(CONFIG_LORAMAC_REGION_EU_868))
```
Comment (jia200x) [inline sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_region.c:48]: as discussed offline, this should be it's own function for every region:
```c
#if (IS_ACTIVE(CONFIG_LORAMAC_REGION_EU_868))
uint8_t gnrc_lorawan_rx1_get_dr_offset(uint8_t dr_up, uint8_t dr_offset)
{
...
}
#endif
```

Otherwise we cannot split these files later

Comment (akshaim) [reply sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_region.c]: Sure.

**Comment by akshaim** [comment]
> Thanks for this!
> Here are some comments.
> 
> Also, could you please add the corresponding Kconfig file to the new symbols?

As discussed offline, shall make it a separate PR .

File: sys/include/net/loramac.h (hunk 6 above)
Comment (jia200x) [reply sys/include/net/loramac.h]: I think I would semtech pkg notation here, since we already configure the pkg with that suffix

File: sys/include/net/loramac.h [hunk 17]
```diff
@@ -105,8 +107,9 @@ extern "C" {
 /**
  * @brief Default LoRaWAN region
  */
-#if !defined(LORAMAC_REGION_EU_868) && !defined(LORAMAC_REGION_IN_865)
-#define LORAMAC_REGION_EU_868                  (1)
+#if !IS_ACTIVE(CONFIG_LORAMAC_REGION_EU_868) \
+    && !IS_ACTIVE(CONFIG_LORAMAC_REGION_IN_865)
+#define CONFIG_LORAMAC_REGION_EU_868           1
```
Comment (jia200x) [inline sys/include/net/loramac.h:112]: Removing the parenthesis should be done in the first commit (130519ade5e157e2437554a279410c4da2c10eaf).

Rationale: This is the first commit that will build (all the others fail because of the extra parenthesis). If someone does `git bisect` and hit those commits, the binary won't compile. Thus, all commits should compile.

File: sys/include/net/loramac.h [hunk 18]
```diff
@@ -210,9 +213,9 @@ extern "C" {
  * @brief   Default second RX window datarate index
  */
 #ifndef LORAMAC_DEFAULT_RX2_DR
-#ifdef LORAMAC_REGION_EU_868
+#if IS_ACTIVE(CONFIG_LORAMAC_REGION_EU_868)
 #define LORAMAC_DEFAULT_RX2_DR                 (LORAMAC_DR_0)
```
Comment (jia200x) [inline sys/include/net/loramac.h:217]: Ditto

File: sys/include/net/loramac.h (hunk 17 above)
Comment (akshaim) [reply sys/include/net/loramac.h:112]: I have used `#Ifdef `for other commits. Just checked. It seems to build just fine. Not sure if I am missing something here

File: examples/gnrc_lorawan/Makefile [hunk 19]
```diff
@@ -57,3 +57,10 @@ CFLAGS += -DLORAMAC_DEFAULT_JOIN_PROCEDURE=LORAMAC_JOIN_OTAA
 CFLAGS += -DLORAMAC_DEFAULT_TX_MODE=LORAMAC_TX_CNF
 
 include $(RIOTBASE)/Makefile.include
+
+# Set region
+ifndef CONFIG_LORAMAC_REGION_EU_868
+  ifndef CONFIG_LORAMAC_REGION_IN_865
+    CFLAGS += -DCONFIG_LORAMAC_REGION_EU_868=1
+  endif
+endif
```
Comment (leandrolanzieri) [inline examples/gnrc_lorawan/Makefile:66]: What about...?
```suggestion
ifeq (,$(CONFIG_LORAMAC_REGION_EU_868)$(CONFIG_LORAMAC_REGION_IN_865))
    CFLAGS += -DCONFIG_LORAMAC_REGION_EU_868=1
endif
```

File: sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_region.c [hunk 20]
```diff
@@ -38,11 +39,24 @@ int gnrc_lorawan_set_dr(gnrc_lorawan_t *mac, uint8_t datarate)
...
 
+#elif (IS_ACTIVE(CONFIG_LORAMAC_REGION_IN_865))
+uint8_t gnrc_lorawan_rx1_get_dr_offset(uint8_t dr_up, uint8_t dr_offset)
+{
+    if (dr_offset < 6) {
+        return (dr_up <= dr_offset) ? 0 : (dr_up - dr_offset) > 5 ? 5 : (dr_up - dr_offset);
+    }
+    else {
+        return (dr_offset + dr_up) > 10 ? 5 : ((dr_offset - 5) + dr_up);
+    }
```
Comment (jia200x) [inline sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_region.c:62]: Last nit pick.
I find this a little bit hard to follow.  Maybe something like this?

```suggestion
    int dr_eff = dr_offset > 5 ? 5 - dr_offset : dr_offset;
    return MIN(5, MAX(0, dr_up - dr_eff));
```

Given:
```
#define MIN(a, b) ((a) < (b) ? (a) : (b))
#define MAX(a, b) ((a) > (b) ? (a) : (b))
```

Comment (akshaim) [reply sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_region.c:62]: Makes sense :+1:

File: examples/gnrc_lorawan/Makefile (hunk 19 above)
Comment (akshaim) [reply examples/gnrc_lorawan/Makefile]: Neat. Thanks for the suggestion :+1:

**Comment by jia200x** [comment]
awesome! Please squash

**Comment by jia200x** [comment]
Murdock found some issues. You should use PRIu32 for uint32_t stuff

**Comment by akshaim** [comment]
> PRIu32

Strange. I 

> Murdock found some issues. You should use PRIu32 for uint32_t stuff

Strange. Did not see this error in "b-l072z-lrwan1" target.

**Comment by aabadie** [comment]
> Did not see this error in "b-l072z-lrwan1" target.

Because you are building with GCC on ARM. The problem is with other toolchains (llvm and gcc on ESP). So for portability reasons, you have to use PRIu32 in printf/DEBUG statements.

**Comment by akshaim** [comment]
> PRIu32

Updated and squashed as its a minor change.

**Comment by akshaim** [comment]
> > Did not see this error in "b-l072z-lrwan1" target.
> 
> Because you are building with GCC on ARM. The problem is with other toolchains (llvm and gcc on ESP). So for portability reasons, you have to use PRIu32 in printf/DEBUG statements.

Thanks for letting me know in detail :+1:

**Review by jia200x — APPROVED** [review]
I cannot test the IN865 region, but I can ensure that the EU868 still works ok. Thanks for this one!
ACK

**Comment by akshaim** [comment]
> I cannot test the IN865 region, but I can ensure that the EU868 still works ok. Thanks for this one!
> ACK

I can ensure IN865 works :)
